src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java

```diff
@@ -0,0 +1,477 @@
+    }
+
+    @Override
+    public void register(IndexRegistry registry)
+    {
+        // index will be available for writes
+        registry.registerIndex(this, StorageAttachedIndexGroup.class, () -> new StorageAttachedIndexGroup(baseCfs));
+    }
+
+    @Override
+    public IndexMetadata getIndexMetadata()
+    {
+        return config;
+    }
+
+    @Override
+    public Callable<?> getInitializationTask()
+    {
+        // New storage-attached indexes will be available for queries after on disk index data are built.
+        // Memtable data will be indexed via flushing triggered by schema change
+        return () -> startInitialBuild(baseCfs).get();
+    }
+
+    private Future<?> startInitialBuild(ColumnFamilyStore baseCfs)
+    {
+        if (baseCfs.indexManager.isIndexQueryable(this))
+        {
+            logger.debug(indexContext.logMessage("Skipping validation and building in initialization task, as pre-join has already made the storage attached index queryable..."));
+        }
+        baseCfs.indexManager.makeIndexQueryable(this, Status.BUILD_SUCCEEDED);
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public Callable<?> getMetadataReloadTask(IndexMetadata indexMetadata)
+    {
+        return null;
+    }
+
+    @Override
+    public Callable<?> getBlockingFlushTask()
+    {
+        return null; // storage-attached indexes are flushed alongside memtable
+    }
+
+    @Override
+    public Callable<?> getInvalidateTask()
+    {
+        return () ->
+        {
+            indexContext.invalidate();
+            return null;
+        };
+    }
+
+    @Override
+    public Callable<?> getPreJoinTask(boolean hadBootstrap)
+    {
+        /*
+         * During bootstrap, streamed SSTable are already built for existing indexes via {@link StorageAttachedIndexBuildingSupport}
+         * from {@link org.apache.cassandra.streaming.StreamReceiveTask.OnCompletionRunnable}.
+         *
+         * For indexes created during bootstrapping, we don't have to block bootstrap for them.
+         */
+
+        return this::startPreJoinTask;
+    }
+
+    private Future<?> startPreJoinTask()
+    {
+        try
+        {
+            if (baseCfs.indexManager.isIndexQueryable(this))
+                logger.debug(indexContext.logMessage("Skipping validation in pre-join task, as the initialization task has already made the index queryable..."));
+
+            baseCfs.indexManager.makeIndexQueryable(this, Status.BUILD_SUCCEEDED);
+        }
+        catch (Throwable t)
+        {
+            logger.error(indexContext.logMessage("Failed in pre-join task!"), t);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Callable<?> getTruncateTask(long truncatedAt)
+    {
+        /*
+         * index files will be removed as part of base sstable lifecycle in
+         * {@link LogTransaction#delete(java.io.File)} asynchronously.
+         */
+        return null;
+    }
```

Comment:
This section of methods above is just taking care of queryability concerns for now, but will be expanded once we bring in the on-disk bits, correct?

Reply:
Correct, there are a number of these methods lying around that will get padded out in the next ticket.